src/unitxt/inference.py

```diff
@@ -31,3 +41,117 @@ def infer(self, dataset):
                 max_new_tokens=self.max_new_tokens,
             )
         ]
+
+
+@dataclass()
+class IbmGenAiInferenceEngineParams:
+    decoding_method: str = None
+    max_new_tokens: Optional[int] = None
+    min_new_tokens: Optional[int] = None
+    random_seed: Optional[int] = None
+    repetition_penalty: Optional[float] = None
+    stop_sequences: Optional[List[str]] = None
+    temperature: Optional[float] = None
+    top_k: Optional[int] = None
+    top_p: Optional[float] = None
+    typical_p: Optional[float] = None
+
+
+class IbmGenAiInferenceEngine(InferenceEngine):
+    label: str = "ibm_genai"
+    model_name: str
+    parameters: IbmGenAiInferenceEngineParams = IbmGenAiInferenceEngineParams()
+
+    def prepare(self):
+        try:
+            from genai import Client, Credentials
+        except ImportError as e:
+            raise ImportError(
+                "Failed to import ibm-genai package. "
+                "Please run 'pip install --upgrade ibm-generative-ai'"
+            ) from e
+
+        api_key_env_var_name = "GENAI_KEY"
+        api_key = os.environ.get(api_key_env_var_name)
+        assert api_key is not None, (
+            f"Error while trying to run IbmGenAiInferenceEngine."
+            f" Please set the environment param '{api_key_env_var_name}'."
+        )
+        api_endpoint = os.environ.get("GENAI_KEY")
+        credentials = Credentials(api_key=api_key, api_endpoint=api_endpoint)
+        self.client = Client(credentials=credentials)
+
+        self._assert_allow_passing_data_to_remote_api(self.label)
+
+    def infer(self, dataset):
+        from genai.schema import TextGenerationParameters
+
+        genai_params = TextGenerationParameters(**self.parameters.__dict__)
+        return list(
+            self.client.text.generation.create(
+                model_id=self.model_name,
+                inputs=[instance["source"] for instance in dataset],
+                parameters=genai_params,
+            )
+        )
+
+
+@dataclass
+class OpenAiInferenceEngineParams:
+    frequency_penalty: Optional[float] = None
+    presence_penalty: Optional[float] = None
+    max_tokens: Optional[int] = None
+    seed: Optional[int] = None
+    stop: Union[Optional[str], List[str]] = None
+    temperature: Optional[float] = None
+    top_p: Optional[float] = None
+
+
+class OpenAiInferenceEngine(InferenceEngine):
+    label: str = "openai"
+    model_name: str
+    parameters: OpenAiInferenceEngineParams = OpenAiInferenceEngineParams()
+
+    def prepare(self):
```

Comment:
Let's use PackageRequirementsMixin to manage imports, similar to how it's done in operators.

class PackageRequirementsMixin(Artifact):
    """Base class used to automatically check for the existence of required python dependencies for an artifact (e.g. Operator or Metric).

    The _requirement list is either a list of required packages
    (e.g. ["torch","sentence_transformers"]) or a dictionary between required packages
    and detailed installation instructions on how how to install each package.
    (e.g. {"torch" : "Install Torch using `pip install torch`", "sentence_transformers" : Install Sentence Transformers using `pip install sentence-transformers`})
    Note that the package names should be specified as they are used in the python import statement for the package.
    """